proposals/0000-unsatisfiable.rst

```diff
@@ -0,0 +1,599 @@
+
+  ghci> let v = reflexiveEq 0 (0 :: Double)
+
+  <interactive>:32:9: warning: [-Wdeferred-type-errors]
+      • Equality is not reflexive on Double
+      • In the expression: reflexiveEq 0 (0 :: Double)
+        In an equation for ‘v’: v = reflexiveEq 0 (0 :: Double)
+  ghci> v
+  False
+
+We would expect evaluation of ``v`` to throw an error, but it does not!  Instead
+it uses the "redundant" method definition from the ``ReflexiveEq Double``
+instance to yield a value.  See `#16249
+<https://gitlab.haskell.org/ghc/ghc/-/issues/16249>`_ and `#18310
+<https://gitlab.haskell.org/ghc/ghc/-/issues/18310>`_ for discussion of this
+issue.
+
+Ideally, it would be possible for the definition of the ``reflexiveEq`` method
+in the ``ReflexiveEq Double`` instance to explicitly make use of the fact that
+the context is inconsistent, rather than giving a bogus definition that is not
+expected to be called.  Then under ``-fdefer-type-errors``, evaluation of the
+bogus evidence for the constraint would be forced, resulting in the expected
+runtime error.
+
+
+Proposed Change Specification
+-----------------------------
+
+The ``GHC.TypeError`` module (the planned new home for ``TypeError`` per `!6066
+<https://gitlab.haskell.org/ghc/ghc/-/merge_requests/6066>`_) is extended with
+the following definitions::
+
+  class Unsatisfiable (e :: ErrorMessage) where
+    unsatisfiable' :: a
+
+  unsatisfiable :: forall e {rep} (a :: TYPE rep). Unsatisfiable e => a
+  unsatisfiable = unsatisfiable' @e (##)
+
+The ``Unsatisfiable`` class and ``unsatisfiable`` function are exported, but the
+``unsatisfiable'`` class method is not.
+
+``Unsatisfiable`` constraints have the following properties:
+
+* During constraint solving, the solver treats ``Unsatisfiable`` constraints
+  like any other class with no instances.  In particular, an ``Unsatisfiable e``
+  Given constraint can solve a corresponding ``Unsatisfiable e`` Wanted
+  constraint, but not ``Unsatisfiable e'`` for some distinct ``e'``.
+
+* At the end of constraint solving:
+
+  - If at least one Given constraint of the form ``Unsatisfiable e`` is present,
+    the constraint solver will automatically solve all Wanted constraints
+    (including any ``Unsatisfiable`` Wanted constraints). The evidence for a
+    Wanted ``w`` consists of a call to ``unsatisfiable @e @w``.
+
+  - Otherwise, if a Wanted constraint of the form ``Unsatisfiable e`` remains
+    unsolved, a type error is reported but the usual "unsolved constraint" error
+    message is replaced by the custom message that results from normalising and
+    rendering the type ``e``.
+
+* An ``Unsatisfiable`` constraint is never automatically generalised.
+
+* If an ``Unsatisfiable`` Given constraint is present during pattern-match coverage
+  checking, the match is trivially regarded as total.
+
+* If a class instance has an ``Unsatisfiable`` Given constraint in the context,
+  it bypasses the functional dependency check.
```

Comment:
If we are going to include this sort of magic (which, I admit, sounds very nice), I would also recommend that not defining a method or an associated type doesn't trigger a warning when `Unsatisfiable` is given.

Reply:
I had misread this paragraph. But I think this is less radical than the alternative section is making it: I can _already_ skip defining methods, if I do so I get a warning (and a runtime error were I to use the missing method)

```
> instance Num Bool

<interactive>:1:10: warning: [-Wmissing-methods]
    • No explicit implementation for
        ‘+’, ‘*’, ‘abs’, ‘signum’, ‘fromInteger’, and (either ‘negate’
                                                              or
                                                              ‘-’)
    • In the instance declaration for ‘Num Bool’

> True + False
*** Exception: <interactive>:1:10-17: No instance nor default method for class operation +
```

As for associated type: when defining an overlapping instance for error message purposes (a very real use-case), I don't have a choice, I _must_ leave the associated type undefined (again, GHC produces a warning). Because type family instances must be non-overlapping (well, the condition is a bit finer than this, but for this use-case it doesn't matter).

I think that this warning should be skipped when `Unsatisfiable` is in the context (I can, of course, deactivate it myself, but it makes sense in the context)

Reply:
@aspiwack thanks, you've convinced me that omitting class methods when there is `Unsatisfiable` in the context is a reasonable thing to do. Ideally the omitted methods (where there is no default either) would be implemented via a call to `unsatisfiable` (rather than an error about the method being undefined).

I hadn't previously thought about associated type families. I think in the absence of [Constrained Type Families](https://arxiv.org/abs/1706.09715), the best we can do is say that we skip the warning for a missing associated type instance if there is an `Unsatisfiable` constraint in the context.

Reply:
> Ideally the omitted methods (where there is no default either) would be implemented via a call to `unsatisfiable` (rather than an error about the method being undefined).

Indeed. Though it doesn't matter very much. If someone ends up changing the behaviour to use `unsatisifiable` in this case, then won't need a proposal.

> I hadn't previously thought about associated type families. I think in the absence of [Constrained Type Families](https://arxiv.org/abs/1706.09715), the best we can do is say that we skip the warning for a missing associated type instance if there is an `Unsatisfiable` constraint in the context.

I think so too.

Reply:
> Indeed. Though it doesn't matter very much. If someone ends up changing the behaviour to use unsatisifiable in this case, then won't need a proposal.

Some committee members are/were vorried about what *precisely* happens with `-fdefer-type-errors`. I feel there is mixed wishes here.